ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -91,6 +94,26 @@ public void initialize(String hostName, int port, List<AmbryHealthReport> ambryH
     }
   }
 
+  @Override
+  public void setReplicaSealedState(ReplicaId replicaId, boolean isSealed) {
+    HelixAdmin helixAdmin = manager.getClusterManagmentTool();
+    InstanceConfig instanceConfig = helixAdmin.getInstanceConfig(clusterName, instanceName);
+    List<String> sealedReplicas = ClusterMapUtils.getSealedReplicas(instanceConfig);
+    if (sealedReplicas == null) {
+      sealedReplicas = new ArrayList<>();
+    }
+    String partitionId = replicaId.getPartitionId().toPathString();
+
+    if (!isSealed) {
+      sealedReplicas.remove(partitionId);
+    }
+    else if (!sealedReplicas.contains(partitionId)) {
```

Comment:
Yup; technically inefficient (since contains() and remove() are linear time ops on lists) but lists should be small enough it doesn't practically matter

Reply:
Yeah. This was due to Helix constraints. Helix supports simple fields (keys to values), list fields (keys to lists), and map fields (keys to maps). We use the map fields for disk to disk metadata information, so all the keys in the map fields are disk names.

Reply:
To make that clear, instance configs have simple fields (key:value) such as the following:
```
   "simpleFields": {
        ...
        "datacenter": "dc1",
        "rackId": "504",
        "sslPort": "15288"
        ...
    }
```

It has map fields (key:map) that are basically disk to disk metadata.
```
  "mapFields": {
        ...
        "/mnt/u001/ambrydata": {
            "Replicas": "637:107374182400,1724:107374182400,442:107374182400,444:107374182400,5:107374182400,649:107374182400,179:107374182400,415:107374182400,",
            "capacityInBytes": "912680550400",
            "diskState": "AVAILABLE"
        },
        "/mnt/u002/ambrydata": {
            "Replicas": "885:107374182400,1677:107374182400,495:107374182400,594:107374182400,544:107374182400,1667:107374182400,1164:107374182400,1353:107374182400,",
            "capacityInBytes": "912680550400",
            "diskState": "AVAILABLE"
        }
        ...
    }
```

Because Helix does not support adding another level of indirection for the map fields, they could only be used for disk to disk metadata.

So we use the list fields (key:list) for other things (currently just the sealed state info):

```
    "listFields": {
        "SEALED": [
            "1751",
            "1057",
            "1128",
            "988",
            "1273",
            "1771",
            "783"
        ]
    }
```